juniper_codegen/src/graphql_union/mod.rs

```diff
@@ -694,9 +700,16 @@ impl VariantDefinition {
         let resolving_code = gen::async_resolving_code(None);
 
         quote! {
-            if type_name == <#ty as ::juniper::GraphQLType<#scalar>>::name(info).unwrap() {
-                let fut = ::juniper::futures::future::ready({ #expr });
-                return #resolving_code;
+            match <#ty as ::juniper::GraphQLType<#scalar>>::name(info) {
+                Some(name) => {
+                    if type_name == name {
+                        let fut = ::juniper::futures::future::ready({ #expr });
+                        return #resolving_code;
+                    }
+                }
+                None => return Box::pin(::juniper::futures::future::ready(
+                    Err(::juniper::FieldError::from("This GraphQLType has no name"))
```

Comment:
@ilslv to DRYing the generated code a bit, let's move such things to helper functions inside `juniper` crate, and reuse them here.